ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1103,24 +1131,37 @@ ReplicationMetrics getReplicationMetrics() {
   }
 
   static class ExchangeMetadataResponse {
-    final Set<StoreKey> missingStoreKeys;
+    // Set of messages missing in the local store.
+    final Set<MessageInfo> missingStoreMessages;
     final FindToken remoteToken;
     final long localLagFromRemoteInBytes;
     final ServerErrorCode serverErrorCode;
 
-    ExchangeMetadataResponse(Set<StoreKey> missingStoreKeys, FindToken remoteToken, long localLagFromRemoteInBytes) {
-      this.missingStoreKeys = missingStoreKeys;
+    ExchangeMetadataResponse(Set<MessageInfo> missingStoreMessages, FindToken remoteToken,
+        long localLagFromRemoteInBytes) {
+      this.missingStoreMessages = missingStoreMessages;
       this.remoteToken = remoteToken;
       this.localLagFromRemoteInBytes = localLagFromRemoteInBytes;
       this.serverErrorCode = ServerErrorCode.No_Error;
     }
 
     ExchangeMetadataResponse(ServerErrorCode errorCode) {
-      missingStoreKeys = null;
-      remoteToken = null;
-      localLagFromRemoteInBytes = -1;
+      this.missingStoreMessages = null;
+      this.remoteToken = null;
+      this.localLagFromRemoteInBytes = -1;
       this.serverErrorCode = errorCode;
     }
+
+    /**
+     * Utility method to extract store keys from messages
+     */
+    static Set<StoreKey> getStoreKeysFromMessages(Set<MessageInfo> messageInfos) {
```

Comment:
According to the usage of this method, looks like it doesn't need to be a static utility method. I think a non-static method like `getMissingStoreKeys` would be more appropriate. Then the code is something like this:
```
return missingStoreMessages == null ? Collections.emptySet() : missingStoreMessages.stream().map(MessageInfo::getStoreKey).collect(Collectors.toSet());
```